### ISSUE TITLE
`polars.DataFrame.filter` panicks starting in polars version `0.20.3` when used with `polars.all_horizontal` expressions containing `polars.exclude` or when used with `polars.fold` 

### ISSUE THREAD

**Issue by MG-MW** [pola-rs/polars#13534]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python  
import polars as pl
from polars.exceptions import PolarsPanicError


df = pl.DataFrame(
    {
        "a": [None, None, None, None],
        "b": [1, 2, None, 1],
        "c": [1, None, None, 1],
    }
)
print(f"Polars version: {pl.__version__}")

print("with all")
df.filter(~pl.all_horizontal(pl.all().is_null()))

print("with exclude")
try:
    df.filter(~pl.all_horizontal(pl.exclude("b").is_null()))  # panicks
except PolarsPanicError:
    print("\texclude panicks")

print("with fold")
try:
    df.filter(
        ~pl.fold(
            acc=pl.lit(True),
            exprs=pl.exclude("b"),
            function=lambda acc, s: acc & s.is_null(),
        )
    )  # panicks
except PolarsPanicError:
    print("\tfold panicks")
```  

Output with version 0.20.3:  
```
Polars version: 0.20.3
with all
with exclude
	exclude panicks
with fold
	fold panicks
thread '<unnamed>' panicked at crates/polars-plan/src/logical_plan/conversion.rs:162:30:
no `selector` expected at this point
thread '<unnamed>' panicked at crates/polars-plan/src/logical_plan/conversion.rs:162:30:
no `selector` expected at this point
```

<details>
<summary>Stack trace for <code>df.filter(~pl.all_horizontal(pl.exclude("b").is_null()))</code></summary>  

```python-traceback  
---------------------------------------------------------------------------
PanicException                            Traceback (most recent call last)
Cell In[4], line 1
----> 1 df.filter(~pl.all_horizontal(pl.exclude(\"b\").is_null()))  # panicks

File ~/.envs/py311/lib/python3.11/site-packages/polars/dataframe/frame.py:4247, in DataFrame.filter(self, *predicates, **constraints)
   4148 def filter(
   4149     self,
   4150     *predicates: (
   (...)
   4157     **constraints: Any,
   4158 ) -> DataFrame:
   4159     \"\"\"
   4160     Filter the rows in the DataFrame based on one or more predicate expressions.
   4161 
   (...)
   4245 
   4246     \"\"\"
-> 4247     return self.lazy().filter(*predicates, **constraints).collect(_eager=True)

File ~/.envs/py311/lib/python3.11/site-packages/polars/lazyframe/frame.py:1749, in LazyFrame.collect(self, type_coercion, predicate_pushdown, projection_pushdown, simplify_expression, slice_pushdown, comm_subplan_elim, comm_subexpr_elim, no_optimization, streaming, background, _eager)
   1746 if background:
   1747     return InProcessQuery(ldf.collect_concurrently())
-> 1749 return wrap_df(ldf.collect())

PanicException: no `selector` expected at this point
```  

</details>  

Output with version 0.20.2:  
```
Polars version: 0.20.2
with all
with exclude
with fold
```

### Issue description

`polars.fold` expressions and  `polars.all_horizontal` expressions containing `polars.exclude` now panick when used in `polars.DataFrame.filter` in polars version `0.20.3`.

### Expected behavior

As in previous versions `polars.DataFrame.filter` should work with fold expressions or `polars.all_horizontal` (and `polars.any_horizontal` etc.) not working on all columns.

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.3
Index type:           UInt32
Platform:             Linux-6.3.1-arch1-1-x86_64-with-glibc2.37
Python:               3.11.3 (main, May  2 2023, 15:34:37) [GCC 12.2.1 20230201]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          2.2.1
connectorx:           0.3.2
deltalake:            <not installed>
fsspec:               2023.9.2
gevent:               <not installed>
hvplot:               <not installed>
matplotlib:           3.8.0
numpy:                1.26.0
openpyxl:             3.1.2
pandas:               2.1.4
pyarrow:              13.0.0
pydantic:             1.10.13
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           2.0.25
xlsx2csv:             0.8.1
xlsxwriter:           3.1.6
```

</details>

**Comment by ritchie46** [comment]
@alexander-beedie can we maybe revert the `exclude` change until this is can expand nested?

**Comment by alexander-beedie** [comment]
> @alexander-beedie can we maybe revert the `exclude` change until this is can expand nested?

No problem; working it into the pending https://github.com/pola-rs/polars/pull/13336 update.

**Comment by ritchie46** [comment]
Thanks!

**Comment by alexander-beedie** [comment]
> Thanks!

FYI, just to confirm: the just-merged #13336 _did_ include the back-out for this 👌 
I'll add this example into the existing test coverage.

**Comment by MG-MW** [comment]
Resolved as  tested in 

- `0.20.4`  
- `0.20.5`  
- `0.20.6`  

Thanks.
